sopt/src/main/scala/dagr/sopt/parsing/ArgTokenCollator.scala

```diff
@@ -51,31 +51,41 @@ class ArgTokenCollator(argTokenizer: ArgTokenizer) extends Iterator[Try[ArgOptio
   /** True if there is another value, false otherwise.  */
   def hasNext: Boolean = nextOption.isDefined
 
-    /** Gets the next token wrapped in a Try.  A failure is returned if the provided [[ArgTokenizer]] returned a failure
-      * or if we could not find an option name ([[ArgOption]] or [[ArgOptionAndValue]] before finding an
-      * option value ([[ArgValue]]).
-      */
+  /** Gets the next token wrapped in a Try.  A failure is returned if the provided [[ArgTokenizer]] returned a failure
+    * or if we could not find an option name ([[ArgOption]] or [[ArgOptionAndValue]] before finding an
+    * option value ([[ArgValue]]).  Once a failure is encountered, no more tokens are returned.
+    */
   def next: Try[ArgOptionAndValues] = {
-    val retVal = nextOption match {
+    nextOption match {
       case None => throw new NoSuchElementException("'next' was called when 'hasNext' is false")
-      case Some(value) => value
+      case Some(Failure(ex)) =>
+        this.nextOption = None
+        Failure(ex)
+      case Some(Success(value)) =>
+        this.advance()
+        Success(value)
     }
-    this.advance()
-    retVal
   }
 
   /** Tries to get the next token that has an option name, and adds any values to `values` if found. */
-  private def nextName(values: ListBuffer[String]): Try[String] = {
-    iterator.next match {
-      case Success(ArgOption(name)) => Success(name)
-      case Success(ArgOptionAndValue(name, value)) => values += value; Success(name)
-      case Success(ArgValue(value)) => Failure(new OptionNameException(s"Illegal option: '$value'"))
-      case Failure(ex) => Failure(ex)
-    }
+  private def nextName(values: mutable.ListBuffer[String]): Try[String] = iterator.head match {
+    case Success(ArgOption(name)) =>
+      iterator.next
+      Success(name)
+    case Success(ArgOptionAndValue(name, value)) =>
+      iterator.next
+      values += value; Success(name)
+    case Success(ArgValue(value)) =>
+      // do not consume the underlying iterator, so that the value is returned by argTokenizer.takeRemaining, otherwise
+      // we would have to keep track of it.  This assumes we do not move past the first failure.
+      Failure(new OptionNameException(s"Illegal option: '$value'"))
+    case Failure(ex) =>
+      iterator.next
+      Failure(ex)
   }
 
   /** Find values with the same option name, may only be ArgValue and ArgOptionAndValue. */
-  private def addValuesWithSameName(name: String, values: ListBuffer[String]): Unit = {
+  private def addValuesWithSameName(name: String, values: mutable.ListBuffer[String]): Unit = {
```

Comment:
It's up to you, but I still find this redundant since a) `ListBuffer` only exists in `mutable` and b) the `Buffer` in the name clearly implies a mutable builder type.